ghost/core/core/server/data/migrations/versions/5.89/2024-07-30-19-51-06-backfill-offer-redemptions.js

```diff
@@ -0,0 +1,62 @@
+// For information on writing migrations, see https://www.notion.so/ghost/Database-migrations-eb5b78c435d741d2b34a582d57c24253
+
+const logging = require('@tryghost/logging');
+const {default: ObjectID} = require('bson-objectid');
+
+// For DML - data changes
+const {createTransactionalMigration} = require('../../utils');
+
+module.exports = createTransactionalMigration(
+    async function up(knex) {
+        // Backfill missing offer redemptions
+        try {
+            // Select all subscriptions that have an `offer_id` but don't have a matching row in the `offer_redemptions` table
+            logging.info('Selecting subscriptions with missing offer redemptions');
+            const result = await knex.raw(`
+                SELECT
+                    mscs.id AS subscription_id,
+                    mscs.offer_id,
+                    mscs.start_date AS created_at,
+                    m.id AS member_id
+                FROM
+                    members_stripe_customers_subscriptions mscs
+                        JOIN 
+                    members_stripe_customers msc ON mscs.customer_id = msc.customer_id
+                        JOIN
+                    members m ON msc.member_id = m.id
+                        LEFT JOIN
+                    offer_redemptions r ON r.subscription_id = mscs.id
+                WHERE
+                    mscs.offer_id IS NOT NULL and r.id IS NULL;
+            `);
+            // knex.raw() returns a different result depending on the database. We need to handle either case
+            let rows = [];
+            // CASE: MySQL returns an array with two elements: the first element is an array of rows, the second element is metadata
+            if (result && result.length > 0 && Array.isArray(result[0])) {
+                rows = result[0];
+            // CASE: SQLite returns an array of rows
+            } else if (result && result.length > 0 && Object.prototype.hasOwnProperty.call(result[0], 'subscription_id')) {
+                rows = result;
+            }
```

Comment:
There's probably a prettier way to do this by checking the `knex.client` information, but this way doesn't make any assumptions about specific clients. It works, but I'm open to suggestions on a cleaner way to do this.

Reply:
Oh awesome, thanks! Will update that shortly...

Reply:
I wish we had a good place to add a checkpoint for things like this. The GH migration checklist doesn't seem right as it could be noisy. Maybe we could consider adding a section like 'platform considerations'? This difference in the return for sqlite vs mysql seems like a big (common) one that we'd want to help ensure we don't miss.

Reply:
Yeah — what do you think about a linting rule to warn you when using `knex.raw()`? There may be other sqlite vs mysql nuances for knex, but this seems like one that would be somewhat common. Not sure if we could detect if you've handled it correctly, but it could at least warn you and point you to `Databaseinfo.isSqlite`, and require the developer to manually acknowledge it by disabling the rule for that line?

Reply:
For now, I've [added a check to the migrations checklist](https://github.com/TryGhost/Ghost/commit/f147167a291a635f041c208da81cbc3aff5bb1b2) to at least ensure we test migrations in both DBs. This was already in the notion doc, but not in the checklist that's added to PRs with migrations.